playwright/playwright.config.ts

```diff
@@ -0,0 +1,29 @@
+import { PlaywrightTestConfig, devices } from '@playwright/test';
+
+const chromium = 'chromium' as const;
+const allSupportedBrowsers = [chromium, 'firefox', 'webkit'] as const;
+const mobileDeviceNames = ['iPhone 11'] as const;
+const platform = process.env.platform ? process.env.platform : 'desktop';
+const debug = !!process.env.debug;
+const browsers =
+  process.env.browsers === 'all' ? allSupportedBrowsers : [chromium];
+
+const config: PlaywrightTestConfig = {
+  use: {
+    headless: !debug,
+  },
+  projects:
+    platform === 'desktop'
+      ? browsers.map(browser => ({
+          name: browser,
+          use: {
+            browserName: browser,
+          },
+        }))
+      : mobileDeviceNames.map(deviceName => ({
+          name: deviceName,
+          use: { ...devices[deviceName] },
```

Comment:
Novice TypeScript question: what's the difference between this and

```suggestion
          use: devices[deviceName],
```

Reply:
Interesting - I just copied it from the [docs](https://playwright.dev/docs/test-configuration); the difference is that this clones the object (ie copy-by-value) and your suggestion references it. This is useful for when objects get mutated further down the line and it causes very confusing behaviour in (seemingly) unrelated places. I wonder if it's in the docs like this so that if people want to override device config they don't mutate the library's copy of it? Not sure!